[flang][Runtime] Add SIGNAL intrinisic

tblah opened the pull request:
The intrinsic is defined as a GNU extension here:
https://gcc.gnu.org/onlinedocs/gfortran/SIGNAL.html

And as an IBM extension here:
https://www.ibm.com/docs/en/xffbg/121.141?topic=procedures-signali-proc-extension

The IBM version provides a compatible subset of the functionality offered by the GNU version. This patch supports most of the GNU features, but not calling SIGNAL as a function. We don't currently support intrinsics being both subroutines and functions and this change seemed too large to be justified by a non-standard intrinsic.

I have not found open source code using this intrinsic. This is needed for a proprietary code base.

tblah committed:
[flang][Runtime] Add SIGNAL intrinisic

The intrinsic is defined as a GNU extension here:
https://gcc.gnu.org/onlinedocs/gfortran/SIGNAL.html

And as an IBM extension here:
https://www.ibm.com/docs/en/xffbg/121.141?topic=procedures-signali-proc-extension

The IBM version provides a compatible subset of the functionality
offered by the GNU version. This patch supports most of the GNU
features, but not calling SIGNAL as a function. We don't currently
support intrinsics being both subroutines AND functions and this changed
seemed too large to be justified by a non-standard intrinsic.

I cannot point to open source code Fortran using this intrinsic. This is
needed for a proprietary code base.

vzakhari commented on flang/lib/Optimizer/Builder/Runtime/Intrinsics.cpp:
```diff
@@ -235,3 +237,66 @@ void fir::runtime::genSystemClock(fir::FirOpBuilder &builder,
+        static_cast<std::int64_t>(reinterpret_cast<std::uintptr_t>(SIG_ERR));
+    mlir::Value sigErr = builder.createIntegerConstant(loc, int64, sigErrVal);
+    returnStatus(sigErr);
+    errno = EINVAL;
```
What effect does this have?

tblah replied:
Here I am trying to emulate what `signal` would do when given an invalid argument. `errno` is a global variable and wouldn't be difficult to declare and read from in the fortran program using signal, so I added this just in case. EINVAL generally means something like "invalid argument".

vzakhari replied:
Hmm, but `errno` is set here during the compilation not during the program execution.  Am I missing something?

tblah replied:
Ahh you are right. Oops!

I will remove the special case for windows. `SIG_IGN` and `SIG_DFL` are in C89 so the Windows implementation of signal may still support them, and if they aren't, presumably the programmer would realize while trying to look up/include the correct integer constants to use.

vzakhari reviewed (commented):
Thanks for the changes!

tblah committed:
Use more accurate pointer type

tblah commented on flang/include/flang/Optimizer/Builder/Runtime/RTBuilder.h:
```diff
@@ -137,6 +137,13 @@ constexpr TypeBuilderFunc getModel<void *>() {
 template <>
+constexpr TypeBuilderFunc getModel<void (*)(int)>() {
+  return [](mlir::MLIRContext *context) -> mlir::Type {
+    return fir::LLVMPointerType::get(context,
```
Thanks for taking a look, yeah I'll update it so that it is easier to read. But I don't think it matters. Since LLVM 17 there has been no support for typed pointers in LLVM IR: https://llvm.org/docs/OpaquePointers.html

kiranchandramohan commented:
> I have not found open source code using this intrinsic. 

The CFL3D code in NASA's repository uses it. https://github.com/nasa/CFL3D/blob/776d5169f34d26ae65701c87038e63b125b7d86a/source/ronnie/main.F#L136


The Windows CI is failing.
```
******************** TEST 'Flang :: Lower/Intrinsics/signal.f90' FAILED ********************
  | Exit Code: 1
  |  
  | Command Output (stdout):
  | --
  | # RUN: at line 2
  | bbc -emit-hlfir -o - C:\ws\src\flang\test\Lower\Intrinsics\signal.f90 \| c:\ws\src\build\bin\filecheck.exe C:\ws\src\flang\test\Lower\Intrinsics\signal.f90
  | # executed command: bbc -emit-hlfir -o - 'C:\ws\src\flang\test\Lower\Intrinsics\signal.f90'
  | # note: command had no output on stdout or stderr
  | # executed command: 'c:\ws\src\build\bin\filecheck.exe' 'C:\ws\src\flang\test\Lower\Intrinsics\signal.f90'
  | # .---command stderr------------
  | # \| C:\ws\src\flang\test\Lower\Intrinsics\signal.f90:52:10: error: CHECK: expected string not found in input
  | # \| ! CHECK: %[[VAL_30:.*]] = arith.constant 1 : i32
  | # \|          ^
  | # \| <stdin>:81:36: note: scanning from here
  | # \|  %c12_i32 = arith.constant 12 : i32
  | # \|                                    ^
  | # \| <stdin>:85:4: note: possible intended match here
  | # \|  %c0_i32 = arith.constant 0 : i32
  | # \|    ^
  | # \|
  | # \| Input file: <stdin>
  | # \| Check file: C:\ws\src\flang\test\Lower\Intrinsics\signal.f90
  | # \|
  | # \| -dump-input=help explains the following input dump.
  | # \|
  | # \| Input was:
  | # \| <<<<<<
  | # \|             .
  | # \|             .
  | # \|             .
  | # \|            76:  fir.global internal @_QMmFsetup_signalsECsigusr1 constant : i32 {
  | # \|            77:  %c10_i32 = arith.constant 10 : i32
  | # \|            78:  fir.has_value %c10_i32 : i32
  | # \|            79:  }
  | # \|            80:  fir.global internal @_QMmFsetup_signalsECsigusr2 constant : i32 {
  | # \|            81:  %c12_i32 = arith.constant 12 : i32
  | # \| check:52'0                                        X error: no match found
  | # \|            82:  fir.has_value %c12_i32 : i32
  | # \| check:52'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  | # \|            83:  }
  | # \| check:52'0     ~~~
  | # \|            84:  fir.global internal @_QMmFsetup_signalsEstat : i32 {
  | # \| check:52'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  | # \|            85:  %c0_i32 = arith.constant 0 : i32
  | # \| check:52'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  | # \| check:52'1        ?                               possible intended match
  | # \|            86:  fir.has_value %c0_i32 : i32
  | # \| check:52'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  | # \|            87:  }
  | # \| check:52'0     ~~~
  | # \|            88:  func.func private @_FortranASignal(i64, !fir.llvm_ptr<() -> ()>) -> i64 attributes {fir.runtime}
  | # \| check:52'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  | # \|            89: }
  | # \| check:52'0     ~~
  | # \| >>>>>>
  | # `-----------------------------
  | # error: command failed with exit status: 1
 

```

tblah commented:
> The CFL3D code in NASA's repository uses it. https://github.com/nasa/CFL3D/blob/776d5169f34d26ae65701c87038e63b125b7d86a/source/ronnie/main.F#L136

Thanks for finding this. Do you know off the top of your head which compiler they are using? The 3 argument version seems to be using a defintion which is not compatible with the GFortran version that I have implemented here (the third argument is an INTENT(OUT) and they are passing a literal -1).

tblah committed:
Remove Windows special case

It turns out SIG_DFL and SIG_IGN are part of the C89 standard so I presume
they are supported on Windows.

tblah commented:
> > The CFL3D code in NASA's repository uses it. https://github.com/nasa/CFL3D/blob/776d5169f34d26ae65701c87038e63b125b7d86a/source/ronnie/main.F#L136
> 
> Thanks for finding this. Do you know off the top of your head which compiler they are using? The 3 argument version seems to be using a defintion which is not compatible with the GFortran version that I have implemented here (the third argument is an INTENT(OUT) and they are passing a literal -1).

CFL3D has an example Makefile using gfortran, but in my own test, if -1 is given as the status argument, gfortran will generate code that segfaults.

It looks like `ifort` has a different definition of the third argument: https://www.intel.com/content/www/us/en/docs/fortran-compiler/developer-guide-reference/2023-0/signal.html. 

I will update to make status INTENT(INOUT) and supporting the ifort flags.

tblah committed:
[flang][runtime] Implement SLEEP intrinsic (#79074)

This intrinsic is a gnu extension. See
https://gcc.gnu.org/onlinedocs/gfortran/SLEEP.html

This intrinsic is used in minighost:
https://github.com/Mantevo/miniGhost/blob/c2102b521568a74862fa5abb074b1fc8041fc222/ref/MG_UTILS.F#L606

tblah commented:
I'm not sure what went wrong with Windows CI, but it probably is not related to this patch. I'll run it again after rebasing.
```
[5379/5382] Generating ../../../../include/flang/ieee_exceptions.f18.mod
[5380/5382] Generating ../../../../incllvm-lit.py: C:\ws\src\llvm\utils\lit\lit\llvm\config.py:57: note: using lit tools: C:\Program Files\Git\usr\bin
llvm-lit.py: C:\ws\src\llvm\utils\lit\lit\llvm\config.py:270: fatal: Could not run process ['C:/ws/src/build/./bin\\llvm-config', '--assertion-mode']
lude/flang/ieee_arithmetic.mod
[5381/5382] Generating ../../../../include/flang/ieee_arithmetic.f18.mod
[5381/5382] Running the Flang regression tests
FAILED: tools/flang/test/CMakeFiles/check-flang
cmd.exe /C "cd /D C:\ws\src\build\tools\flang\test && C:\Python39\python.exe C:/ws/src/build/./bin/llvm-lit.py -v --xunit-xml-output C:/ws/src/build/test-results.xml --timeout=1200 --time-tests --param flang_site_config=C:/ws/src/build/tools/flang/test/lit.site.cfg.py C:/ws/src/build/tools/flang/test"
ninja: build stopped: subcommand failed.
```

yi-wu-arm commented:
I've built on Windows MSVC and pass all tests.

tblah commented:
I resolved conflicts and merged manually. See https://github.com/llvm/llvm-project/commit/afa52de9f6def6a0de962401f9a6b34925f7010e
